dataclients/kubernetes/definitions/routegroupvalidator.go

```diff
@@ -0,0 +1,330 @@
+	for _, b := range rg.Backends {
+		if backends[b.Name] {
+			return backendsWithDuplicateName(b.Name)
+		}
+
+		backends[b.Name] = true
+		if err := b.validate(); err != nil {
+			return invalidBackend(b.Name, err)
+		}
+	}
+
+	hasDefault := len(rg.DefaultBackends) > 0
+	if err := rg.DefaultBackends.validate(backends); err != nil {
+		return err
+	}
+
+	if !hasDefault && len(rg.Routes) == 0 {
+		return errMissingBackendReference
+	}
+
+	for i, r := range rg.Routes {
+		if err := r.validate(hasDefault, backends); err != nil {
+			return invalidRoute(i, err)
+		}
+	}
+
+	return nil
+}
+
+// TODO: we need to pass namespace/name in all errors
+func (r *RouteSpec) validate(hasDefault bool, backends map[string]bool) error {
+	if r == nil {
+		return errInvalidRouteSpec
+	}
+
+	if !hasDefault && len(r.Backends) == 0 {
+		return errMissingBackendReference
+	}
+
+	if err := r.Backends.validate(backends); err != nil {
+		return err
+	}
+
+	if r.Path != "" && r.PathSubtree != "" {
+		return errBothPathAndPathSubtree
+	}
+
+	if hasEmpty(r.Predicates) {
+		return errInvalidPredicate
+	}
+
+	if hasEmpty(r.Filters) {
+		return errInvalidFilter
+	}
+
+	if hasEmpty(r.Methods) {
+		return errInvalidMethod
+	}
+
+	return nil
+}
+
+// TODO: we need to pass namespace/name in all errors
+func (r *RouteGroupItem) validate() error {
+	// has metadata and name:
+	if r == nil || validate(r.Metadata) != nil {
+		return errRouteGroupWithoutName
+	}
+
+	// has spec:
+	if r.Spec == nil {
+		return routeGroupError(r.Metadata, errRouteGroupWithoutSpec)
+	}
+
+	if err := r.Spec.validate(); err != nil {
+		return routeGroupError(r.Metadata, err)
+	}
+
+	return nil
+}
+
+// ParseRouteGroupsJSON parses a json list of RouteGroups into RouteGroupList
+func ParseRouteGroupsJSON(d []byte) (RouteGroupList, error) {
+	var rl RouteGroupList
+	err := json.Unmarshal(d, &rl)
+	return rl, err
+}
+
+// ParseRouteGroupsYAML parses a YAML list of RouteGroups into RouteGroupList
+func ParseRouteGroupsYAML(d []byte) (RouteGroupList, error) {
+	var rl RouteGroupList
+	err := yaml.Unmarshal(d, &rl)
+	return rl, err
+}
+
+type RouteGroupValidator struct {
+	FilterRegistry filters.Registry
+}
+
+func (rgv *RouteGroupValidator) Validate(item *RouteGroupItem) error {
+	err := validateRouteGroup(item)
+
+	if rgv.FilterRegistry != nil {
+		nerr := validateRouteGroupFilters(item, rgv.FilterRegistry)
+		if nerr != nil {
+			err = errors.Wrap(err, nerr.Error())
+		}
+	}
+
+	return err
+}
+
+// validateRouteGroup validates a RouteGroupItem
+func validateRouteGroup(rg *RouteGroupItem) error {
+	return rg.validate()
+}
+
+func validateRouteGroupFilters(rg *RouteGroupItem, fr filters.Registry) error {
+	// basic for now
+	for _, r := range rg.Spec.Routes {
+		for _, f := range r.Filters {
+			parsedFilter, err := eskip.ParseFilters(f)
+			if err != nil {
+				return err
+			}
+			if _, ok := fr[parsedFilter[0].Name]; !ok {
+				return errors.Errorf("filter %s not found", parsedFilter[0].Name)
```

Comment:
%s -> %q

Reply:
https://pkg.go.dev/github.com/pkg/errors#Errorf

Do we want stacktrace here? Maybe it should fmt.Errorf